cluster-autoscaler/core/scaleup/orchestrator/executor.go

```diff
@@ -0,0 +1,240 @@
+	scaleUpInfos []nodegroupset.ScaleUpInfo,
+	nodeInfos map[string]*schedulerframework.NodeInfo,
+	now time.Time,
+) (errors.AutoscalerError, []cloudprovider.NodeGroup) {
+	options := e.autoscalingContext.AutoscalingOptions
+	if options.ParallelScaleUp {
+		return e.executeScaleUpsParallel(scaleUpInfos, nodeInfos, now)
+	}
+	return e.executeScaleUpsSync(scaleUpInfos, nodeInfos, now)
+}
+
+func (e *scaleUpExecutor) executeScaleUpsSync(
+	scaleUpInfos []nodegroupset.ScaleUpInfo,
+	nodeInfos map[string]*schedulerframework.NodeInfo,
+	now time.Time,
+) (errors.AutoscalerError, []cloudprovider.NodeGroup) {
+	availableGPUTypes := e.autoscalingContext.CloudProvider.GetAvailableGPUTypes()
+	for _, scaleUpInfo := range scaleUpInfos {
+		nodeInfo, ok := nodeInfos[scaleUpInfo.Group.Id()]
+		if !ok {
+			klog.Errorf("ExecuteScaleUp: failed to get node info for node group %s", scaleUpInfo.Group.Id())
+			continue
+		}
+		if aErr := e.executeScaleUp(scaleUpInfo, nodeInfo, availableGPUTypes, now); aErr != nil {
+			return aErr, []cloudprovider.NodeGroup{scaleUpInfo.Group}
+		}
+	}
+	return nil, nil
+}
+
+func (e *scaleUpExecutor) executeScaleUpsParallel(
+	scaleUpInfos []nodegroupset.ScaleUpInfo,
+	nodeInfos map[string]*schedulerframework.NodeInfo,
+	now time.Time,
+) (errors.AutoscalerError, []cloudprovider.NodeGroup) {
+	if err := assertUniqueNodeGroups(scaleUpInfos); err != nil {
+		return err, extractNodeGroups(scaleUpInfos)
+	}
+	type errResult struct {
+		err  errors.AutoscalerError
+		info *nodegroupset.ScaleUpInfo
+	}
+	scaleUpsLen := len(scaleUpInfos)
+	errResults := make(chan errResult, scaleUpsLen)
+	var wg sync.WaitGroup
+	wg.Add(scaleUpsLen)
+	availableGPUTypes := e.autoscalingContext.CloudProvider.GetAvailableGPUTypes()
+	for _, scaleUpInfo := range scaleUpInfos {
+		go func(info nodegroupset.ScaleUpInfo) {
+			defer wg.Done()
+			nodeInfo, ok := nodeInfos[info.Group.Id()]
+			if !ok {
+				klog.Errorf("ExecuteScaleUp: failed to get node info for node group %s", info.Group.Id())
+				return
+			}
+			if aErr := e.executeScaleUp(info, nodeInfo, availableGPUTypes, now); aErr != nil {
+				errResults <- errResult{err: aErr, info: &info}
+			}
+		}(scaleUpInfo)
+	}
+	wg.Wait()
+	close(errResults)
+	var results []errResult
+	for err := range errResults {
+		results = append(results, err)
+	}
+	if len(results) > 0 {
+		failedNodeGroups := make([]cloudprovider.NodeGroup, len(results))
+		scaleUpErrors := make([]errors.AutoscalerError, len(results))
+		for i, result := range results {
+			failedNodeGroups[i] = result.info.Group
+			scaleUpErrors[i] = result.err
+		}
+		return combineConcurrentScaleUpErrors(scaleUpErrors), failedNodeGroups
+	}
+	return nil, nil
+}
+
+func (e *scaleUpExecutor) executeScaleUp(
+	info nodegroupset.ScaleUpInfo,
+	nodeInfo *schedulerframework.NodeInfo,
+	availableGPUTypes map[string]struct{},
+	now time.Time,
+) errors.AutoscalerError {
+	gpuConfig := e.autoscalingContext.CloudProvider.GetNodeGpuConfig(nodeInfo.Node())
```

Comment:
This function call a platitude of other objects, did we check that all of them support concurrent calls? 
I did my own quick check and didn't found any obvious issues, but wanted to confirm as any potential issues here will be hard to debug.

Reply:
Good point but the more critical line is [info.Group.IncreaseSize(increase)](https://github.com/pmendelski/autoscaler/blob/934296343a7f8113c49dac9472603023f303f0e1/cluster-autoscaler/core/scaleup/orchestrator/executor.go#L151) as it modifies the state. I did a thorough analysis and finished with a conclusion that not all cloudProviders are ready for concurrent operations. That's why it's a flag and the decision is left to the user. Although I've just added a warning message to flag description.